<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.550">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Ronald Visser">
<meta name="dcterms.date" content="2024-03-05">

<title>Dendrochronology and R - R-packages for dendrochronology</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/cookie-consent/cookie-consent.js"></script>
<link href="site_libs/cookie-consent/cookie-consent.css" rel="stylesheet">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-LLXF7RMDPD"></script>

<script type="text/plain" cookie-consent="tracking">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-LLXF7RMDPD', { 'anonymize_ip': true});
</script>

<script type="text/javascript" charset="UTF-8">
document.addEventListener('DOMContentLoaded', function () {
cookieconsent.run({
  "notice_banner_type":"simple",
  "consent_type":"implied",
  "palette":"light",
  "language":"en",
  "page_load_consent_levels":["strictly-necessary","functionality","tracking","targeting"],
  "notice_banner_reject_button_hide":false,
  "preferences_center_close_button_hide":false,
  "website_name":""
  ,
"language":"en"
  });
});
</script> 
  


</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Dendrochronology and R</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="./index.html" aria-current="page"> 
<span class="menu-text">R-packages for dendrochronology</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html"> 
<span class="menu-text">About this website</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-bi-github" role="button" data-bs-toggle="dropdown" aria-expanded="false">
      <i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-bi-github">    
        <li>
    <a class="dropdown-item" href="https://github.com/RonaldVisser/Dendro_R">
 <span class="dropdown-text">Source code</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://github.com/RonaldVisser/Dendro_R/issues">
 <span class="dropdown-text">Report a Bug</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://akademienl.social/@RonaldVisser"> <i class="bi bi-mastodon" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">R-packages for dendrochronology</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Ronald Visser <a href="mailto:r.m.visser@saxion.nl" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0001-6966-1729" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Saxion University of Applied Sciences
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">March 5, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="dendrochronology-and-r" class="level1">
<h1>Dendrochronology and R</h1>
<p>This site lists R-packages used in dendrochronology or tree-ring research. For more general information on dendrochronology: <a href="https://www.dendrohub.com/">DendroHub</a> is an excellent website (maintained by Joe Buck). This overview of R-packages might not be complete, and if you miss anything, please add these using a <a href="https://github.com/RonaldVisser/Dendro_R/pulls">pull request in Github</a> or contact me.</p>
<p>If you want to start with <a href="https://www.r-project.org/">R</a>, there are several options:</p>
<ul>
<li><a href="https://opendendro.org/">OpenDendro</a> is a great resource for those who want to learn about dplR (or dplPy). There are some wonderful tutorials.</li>
<li><a href="https://software-carpentry.org/lessons/">Software Carpentry</a> is for teaching basic lab skills for research computing and is good for beginners.</li>
<li><a href="https://datacarpentry.org/lessons/">Data carpentry</a> helps you to gain foundational skills needed to work effectively and reproducibly with data and code.</li>
<li><a href="https://r4ds.hadley.nz/">R for Data Science (2e)</a> is an extensive and accessible book on general R and doing data science using R.</li>
<li><a href="https://github.com/benmarwick/ctv-archaeology">CTV-Archaeology</a> is a more general list maintained by Ben Marwick on possibilities and libraries in R, mostly aimed at Archaeology.</li>
</ul>
<p>I would recommend working with an IDE such as <a href="https://posit.co/download/rstudio-desktop/">RStudio</a> and always use <a href="https://r4ds.had.co.nz/workflow-projects.html">projects</a> in your workflow!</p>
</section>
<section id="list-of-r-packages-for-dendrochronology" class="level1">
<h1>List of R-packages for dendrochronology</h1>
<table class="table">
<tbody>
<tr class="odd">
<td><h2 id="package-name" class="anchored" data-anchor-id="list-of-r-packages-for-dendrochronology">Package name</h2></td>
<td><h2 id="description1" class="anchored">Description<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></h2></td>
<td><h2 id="links" class="anchored">Links</h2></td>
<td><h2 id="github" class="anchored">Github</h2></td>
</tr>
<tr class="even">
<td>BIOdry</td>
<td>Multilevel ecological data series (MEDS) are sequences of observations ordered according to temporal/spatial hierarchies that are defined by sample designs, with sample variability confined to ecological factors. Dendroclimatic MEDS of tree rings and climate are modeled into normalized fluctuations of tree growth and aridity <span class="citation" data-cites="lara2018">(<a href="#ref-lara2018" role="doc-biblioref">Lara, Bogino and Bravo 2018</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/BIOdry">CRAN</a></td>
<td><p><a href="https://github.com/cran/BIOdry">GitHub</a></p>
<p>(read only)</p></td>
</tr>
<tr class="odd">
<td>burnr</td>
<td>Basic tools to analyze forest fire history data (e.g.&nbsp;FHX) in R. This is designed for power users and projects with special needs <span class="citation" data-cites="malevich2018">(<a href="#ref-malevich2018" role="doc-biblioref">Malevich, Guiterman and Margolis 2018</a>)</span>.</td>
<td><p><a href="https://cran.r-project.org/web/packages/burnr/index.html">CRAN</a></p>
<p><a href="https://ltrr-arizona-edu.github.io/burnr/">Website</a></p></td>
<td><a href="https://github.com/ltrr-arizona-edu/burnr">GitHub</a></td>
</tr>
<tr class="even">
<td>dfoliatR</td>
<td>Tools to identify, quantify, analyze, and visualize growth suppression events in tree rings that are often produced by insect defoliation <span class="citation" data-cites="guiterman2020">(<a href="#ref-guiterman2020" role="doc-biblioref">Guiterman, Lynch and Axelson 2020</a>)</span>.</td>
<td><p><a href="https://cran.r-project.org/web/packages/dfoliatR/index.html">CRAN</a></p>
<p><a href="https://chguiterman.github.io/dfoliatR/">Website</a></p></td>
<td><a href="https://github.com/chguiterman/dfoliatR/">Github</a></td>
</tr>
<tr class="odd">
<td>dendRoAnalyst</td>
<td>A package to analyse (daily) statistics of dendrometer data, including the daily amplitude of tree growth. Various approaches can be applied to separate radial growth from daily cyclic shrinkage and expansion due to uptake and loss of stem water <span class="citation" data-cites="aryal2020">(<a href="#ref-aryal2020" role="doc-biblioref">Aryal et al. 2020</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/dendRoAnalyst/index.html">CRAN</a></td>
<td><p><a href="https://github.com/cran/dendRoAnalyst">GitHub</a></p>
<p>(read only)</p></td>
</tr>
<tr class="even">
<td>dendroNetwork</td>
<td>dendroNetwork is a package to create dendrochronological networks for gaining insight into provenance or other patterns based on the statistical relations between tree ring curves. The networks can be used to visualise and understand the relations between tree-ring curves <span class="citation" data-cites="visser2024">(<a href="#ref-visser2024" role="doc-biblioref">Visser 2024</a>)</span>.</td>
<td></td>
<td><a href="https://github.com/RonaldVisser/dendroNetwork">GitHub</a></td>
</tr>
<tr class="odd">
<td>dendroSync</td>
<td>Provides functions for the calculation and plotting of synchrony in tree growth from tree-ring width chronologies (TRW index). It combines variance-covariance (VCOV) mixed modelling with functions that quantify the degree to which the TRW chronologies contain a common temporal signal. It also implements temporal trends in spatial synchrony using a moving window. These methods can also be used with other kind of ecological variables that have temporal autocorrelation corrected <span class="citation" data-cites="alday2018">(<a href="#ref-alday2018" role="doc-biblioref">Alday et al. 2018</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/DendroSync/index.html">CRAN</a></td>
<td><a href="https://github.com/cran/DendroSync">GitHub</a></td>
</tr>
<tr class="even">
<td>dendroTools</td>
<td>Provides novel dendroclimatological methods, primarily used by the Tree-ring research community. There are four core functions. The first one is daily_response(), which finds the optimal sequence of days that are related to one or more tree-ring proxy records. Similar function is daily_response_seascorr(), which implements partial correlations in the analysis of daily response functions. For the enthusiast of monthly data, there is monthly_response() function. The last core function is compare_methods(), which effectively compares several linear and nonlinear regression algorithms on the task of climate reconstruction <span class="citation" data-cites="jevsenak2018">(<a href="#ref-jevsenak2018" role="doc-biblioref">Jevšenak and Levanič 2018</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/dendroTools/index.html">CRAN</a></td>
<td><a href="https://github.com/jernejjevsenak/dendroTools" class="uri">GitHub</a></td>
</tr>
<tr class="odd">
<td>detrendeR</td>
<td>A Graphical User Interface (GUI) to import, save, detrend and perform standard tree-ring analyses. The interactive detrending allows the user to check how well the detrending curve fits each time-series and change it when needed <span class="citation" data-cites="campelo2012">(<a href="#ref-campelo2012" role="doc-biblioref">Campelo, García-González and Nabais 2012</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/detrendeR/index.html">CRAN</a></td>
<td><p><a href="https://github.com/cran/detrendeR" class="uri">GitHub</a></p>
<p>(read only)</p></td>
</tr>
<tr class="even">
<td>dplR</td>
<td>General package for tree-ring analyses. It is possible to perform tree-ring analyses such as detrending, chronology building, and cross dating. The sotware reads and writes standard file formats used in dendrochronology <span class="citation" data-cites="bunn2008 bunn2010">(<a href="#ref-bunn2008" role="doc-biblioref">Bunn 2008</a>; <a href="#ref-bunn2010" role="doc-biblioref">Bunn 2010</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/dplR/index.html">CRAN</a></td>
<td><a href="https://github.com/AndyBunn/dplR" class="uri">GItHub</a></td>
</tr>
<tr class="odd">
<td>fellingdateR</td>
<td>This R-package offers a set of functions that can assist you to infer felling date estimates from dated tree-ring series. The presence of (partially) preserved sapwood or waney edge allows to estimate a range for the actual felling date, for individual series as well as for a group of timbers. Furthermore, an additional function provides a tool to sum sapwood probability distributions, comparable to ‘summed probability densities’ commonly applied to sets of radiocarbon (<sup>14</sup>C) dates <span class="citation" data-cites="haneca2023">(<a href="#ref-haneca2023" role="doc-biblioref">Haneca 2023</a>)</span>.</td>
<td></td>
<td><a href="https://github.com/hanecakr/fellingDateR" class="uri">GitHub</a></td>
</tr>
<tr class="even">
<td>measuRing</td>
<td>Identification of ring borders on scanned image sections from dendrochronological samples. Processing of image reflectances to produce gray matrices and time series of smoothed gray values. Luminance data is plotted on segmented images for users to perform both: visual identification of ring borders or control of automatic detection. Routines to visually include/exclude ring borders on the R graphical devices, or automatically detect ring borders using a linear detection algorithm. This algorithm detects ring borders according to positive/negative extreme values in the smoothed time-series of gray values. Most of the in-package routines can be recursively implemented using the multiDetect() function <span class="citation" data-cites="lara2015">(<a href="#ref-lara2015" role="doc-biblioref">Lara, Bravo and Sierra 2015</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/measuRing/index.html">CRAN</a></td>
<td></td>
</tr>
<tr class="odd">
<td>MtreeRing<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a></td>
<td><p>A Shiny Application for Automatic Measurements of Tree-Ring Widths on Digital Images</p>
<p>Use morphological image processing and edge detection algorithms to automatically measure tree ring widths on digital images. Users can also manually mark tree rings on species with complex anatomical structures. The arcs of inner-rings and angles of successive inclined ring boundaries are used to correct ring-width series. The package provides a Shiny-based application, allowing R beginners to easily analyze tree ring images and export ring-width series in standard file formats <span class="citation" data-cites="shi2019">(<a href="#ref-shi2019" role="doc-biblioref">Shi et al. 2019</a>)</span>.</p></td>
<td><a href="https://cran.r-project.org/web/packages/MtreeRing/index.html">CRAN</a></td>
<td><a href="https://github.com/ropensci/MtreeRing/" class="uri">GitHub</a></td>
</tr>
<tr class="even">
<td>pointRes</td>
<td>Functions to calculate and plot event and pointer years as well as resilience indices. Designed for dendroecological applications, but also suitable to analyze patterns in other ecological time series <span class="citation" data-cites="vandermaaten-theunissen2015">(<a href="#ref-vandermaaten-theunissen2015" role="doc-biblioref">Maaten-Theunissen, Maaten and Bouriaud 2015</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/pointRes/index.html">CRAN</a></td>
<td><p><a href="https://github.com/cran/pointRes" class="uri">GitHub</a></p>
<p>(read only)</p></td>
</tr>
<tr class="odd">
<td>RAPTOR</td>
<td>Performs wood cell anatomical data analyses on spatially explicit xylem (tracheids) datasets derived from thin sections of woody tissue. The package includes functions for visualisation, detection and alignment of continuous tracheid radial file (defined as rows) and individual tracheid position within an annual ring of coniferous species <span class="citation" data-cites="peters2018">(<a href="#ref-peters2018" role="doc-biblioref">Peters et al. 2018</a>)</span>.</td>
<td><a href="https://github.com/the-hull/RAPTOR">CRAN</a></td>
<td><a href="https://github.com/the-hull/RAPTOR">GitHub</a></td>
</tr>
<tr class="even">
<td>RingdateR</td>
<td>RingdateR is a graphical and statistical tool for crossdating annually resolved time series such as tree rings or mollusc, coral and otolith growth increments <span class="citation" data-cites="reynolds2021">(<a href="#ref-reynolds2021" role="doc-biblioref">Reynolds, Edge and Black 2021</a>)</span>.</td>
<td><a href="https://ringdater.github.io/ringdater/index.html">Website</a></td>
<td><a href="https://github.com/ringdater/ringdater_pkg" class="uri">GitHub</a></td>
</tr>
<tr class="odd">
<td>SapwoodModel</td>
<td>This software package fits a model to sapwood data from scots pine (Pinus sylvestris L.) using heartwood data and tree ring width using a nonlinear model. Three models are implemented <span class="citation" data-cites="edvardsson2022">(<a href="#ref-edvardsson2022" role="doc-biblioref">Edvardsson et al. 2022</a>)</span>.</td>
<td><a href="https://kar26.shinyapps.io/SapwoodModel/">Shiny</a></td>
<td><a href="https://github.com/karirogg/SapwoodModel">GitHub</a></td>
</tr>
<tr class="even">
<td>tracheideR</td>
<td>Contains functions to standardize tracheid profiles using the traditional method (Vaganov) and a new method to standardize tracheidograms based on the relative position of tracheids within tree rings <span class="citation" data-cites="campelo2016">(<a href="#ref-campelo2016" role="doc-biblioref">Campelo et al. 2016</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/tracheideR/index.html">CRAN</a></td>
<td><p><a href="https://github.com/cran/tracheideR">GitHub</a></p>
<p>(read only)</p></td>
</tr>
<tr class="odd">
<td>TRADER</td>
<td>Tree Ring Analysis of Disturbance Events in R (TRADER) package provides only one way for disturbance reconstruction from tree-ring data <span class="citation" data-cites="altman2014">(<a href="#ref-altman2014" role="doc-biblioref">Altman et al. 2014</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/TRADER/index.html">CRAN</a></td>
<td><a href="https://github.com/pavel-fibich/TRADER" class="uri">GitHub</a></td>
</tr>
<tr class="even">
<td>treeclim</td>
<td>Bootstrapped response and correlation functions, seasonal correlations and evaluation of reconstruction skills for use in dendroclimatology and dendroecology <span class="citation" data-cites="zang2015">(<a href="#ref-zang2015" role="doc-biblioref">Zang and Biondi 2015</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/treeclim/index.html">CRAN</a></td>
<td><a href="https://github.com/cszang/treeclim">GitHub</a></td>
</tr>
<tr class="odd">
<td>ttprocessing</td>
<td>For processing the Tree Talker data. It implements the procedures described in the TT+ manual. It allows for converting the raw data into physical quantities and calculating the sap flow from raw temperatures recorded by the probes. It provides some plotting functions and a graphical user interface <span class="citation" data-cites="kabala2022 kabala2024">(<a href="#ref-kabala2022" role="doc-biblioref">Kabala, Niccoli and Battipaglia 2022</a>; <a href="#ref-kabala2024" role="doc-biblioref">Kabala, Niccoli and Battipaglia 2024</a>)</span>.</td>
<td></td>
<td><a href="https://github.com/jpkabala96/ttprocessing">GitHub</a></td>
</tr>
<tr class="even">
<td>WIAD</td>
<td>wiad: Wood Image Analysis and Dataset provides a user-friendly interface to digitize wood imagery such as tree ring scans. The package offers a web-based application that the user can run locally from their computers to analyze wood characteristics <span class="citation" data-cites="rademacher">(<a href="#ref-rademacher" role="doc-biblioref">Rademacher et al. n.d.</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/wiad/index.html">CRAN</a></td>
<td><a href="https://github.com/bnasr/wiad/">GitHub</a></td>
</tr>
<tr class="odd">
<td>xRing</td>
<td>Identify and Measure Tree Rings on X-Ray Micro-Density Profiles. This package contains functions to identify tree-ring borders based on X-ray micro-density profiles and a Graphical User Interface (GUI) to visualize density profiles and correct tree-ring borders <span class="citation" data-cites="campelo2019">(<a href="#ref-campelo2019" role="doc-biblioref">Campelo, Mayer and Grabner 2019</a>)</span>.</td>
<td><a href="https://cran.r-project.org/web/packages/xRing/index.html">CRAN</a></td>
<td><p><a href="https://github.com/cran/xRing" class="uri">GitHub</a></p>
<p>(read only)</p></td>
</tr>
</tbody>
</table>
</section>
<section id="references" class="level1">




</section>


<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-line-spacing="2" role="list">
<div id="ref-alday2018" class="csl-entry" role="listitem">
Alday, JG, Shestakova, TA, Resco de Dios, V and Voltas, J. 2018 DendroSync: An r package to unravel synchrony patterns in tree-ring networks. <em>Dendrochronologia</em> 47: 17–22. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2017.12.003">10.1016/j.dendro.2017.12.003</a>.
</div>
<div id="ref-altman2014" class="csl-entry" role="listitem">
Altman, J, Fibich, P, Dolezal, J and Aakala, T. 2014 TRADER: A package for tree ring analysis of disturbance events in r. <em>Dendrochronologia</em> 32(2): 107–112. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2014.01.004">10.1016/j.dendro.2014.01.004</a>.
</div>
<div id="ref-aryal2020" class="csl-entry" role="listitem">
Aryal, S, Häusser, M, Grießinger, J, Fan, Z and Bräuning, A. 2020 <span>“</span>dendRoAnalyst<span>”</span>: A tool for processing and analysing dendrometer data. <em>Dendrochronologia</em> 64: 125772. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2020.125772">10.1016/j.dendro.2020.125772</a>.
</div>
<div id="ref-bunn2008" class="csl-entry" role="listitem">
Bunn, AG. 2008 A dendrochronology program library in r (dplR). <em>Dendrochronologia</em> 26(2): 115124. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2008.01.002">https://doi.org/10.1016/j.dendro.2008.01.002</a>.
</div>
<div id="ref-bunn2010" class="csl-entry" role="listitem">
Bunn, AG. 2010 Statistical and visual crossdating in r using the dplR library. <em>Dendrochronologia</em> 28(4): 251–258. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2009.12.001">10.1016/j.dendro.2009.12.001</a>.
</div>
<div id="ref-campelo2012" class="csl-entry" role="listitem">
Campelo, F, García-González, I and Nabais, C. 2012 detrendeR - a graphical user interface to process and visualize tree-ring data using r. <em>Dendrochronologia</em> 30: 57–60. DOI: https://doi.org/<a href="https://doi.org/16/j.dendro.2011.01.010">16/j.dendro.2011.01.010</a>.
</div>
<div id="ref-campelo2019" class="csl-entry" role="listitem">
Campelo, F, Mayer, K and Grabner, M. 2019 xRing<span></span>An R package to identify and measure tree-ring features using X-ray microdensity profiles. <em>Dendrochronologia</em> 53: 17–21. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2018.11.002">10.1016/j.dendro.2018.11.002</a>.
</div>
<div id="ref-campelo2016" class="csl-entry" role="listitem">
Campelo, F, Nabais, C, Carvalho, A and Vieira, J. 2016 tracheideR<span></span>an r package to standardize tracheidograms. <em>Dendrochronologia</em> 37: 64–68. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2015.12.006">10.1016/j.dendro.2015.12.006</a>.
</div>
<div id="ref-edvardsson2022" class="csl-entry" role="listitem">
Edvardsson, J, Rögnvaldsson, K, Helgadóttir, EÞ, Linderson, H and Hrafnkelsson, B. 2022 A statistical model for the prediction of the number of sapwood rings in Scots pine (Pinus sylvestris L.). <em>Dendrochronologia</em> 74: 125963. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2022.125963">10.1016/j.dendro.2022.125963</a>.
</div>
<div id="ref-garcía-hidalgo2021" class="csl-entry" role="listitem">
García-Hidalgo, M, García-Pedrero, ÁM, Caetano-Sánchez, C, Gómez-España, M, Lillo-Saavedra, M and Olano, JM. 2021 ρ-MtreeRing: A Graphical User Interface for X-ray Microdensity Analysis. <em>Forests</em> 12(10): 1405. DOI: https://doi.org/<a href="https://doi.org/10.3390/f12101405">10.3390/f12101405</a>.
</div>
<div id="ref-guiterman2020" class="csl-entry" role="listitem">
Guiterman, CH, Lynch, AM and Axelson, JN. 2020 dfoliatR: An R package for detection and analysis of insect defoliation signals in tree rings. <em>Dendrochronologia</em> 63: 125750. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2020.125750">10.1016/j.dendro.2020.125750</a>.
</div>
<div id="ref-haneca2023" class="csl-entry" role="listitem">
Haneca, K. 2023. <em><a href="https://github.com/hanecakr/fellingDateR">fellingdateR: An r package to facilitate the organisation, analysis and interpretation of tree-ring data from wooden heritage objects and structures</a></em>.
</div>
<div id="ref-jevsenak2018" class="csl-entry" role="listitem">
Jevšenak, J and Levanič, T. 2018 dendroTools: R package for studying linear and nonlinear responses between tree-rings and daily environmental data. <em>Dendrochronologia</em> 48: 32–39. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2018.01.005">10.1016/j.dendro.2018.01.005</a>.
</div>
<div id="ref-kabala2022" class="csl-entry" role="listitem">
Kabala, JP, Niccoli, F and Battipaglia, G. 2022 <em>2022 IEEE workshop on metrology for agriculture and forestry (MetroAgriFor)</em>. In: November 2022. pp. 70–74. DOI: https://doi.org/<a href="https://doi.org/10.1109/MetroAgriFor55389.2022.9964512">10.1109/MetroAgriFor55389.2022.9964512</a>.
</div>
<div id="ref-kabala2024" class="csl-entry" role="listitem">
Kabala, JP, Niccoli, F and Battipaglia, G. 2024 Update to ttprocessing: The r-package to handle the TreeTalker monitoring data. <em>Dendrochronologia</em> 84: 126167. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2024.126167">10.1016/j.dendro.2024.126167</a>.
</div>
<div id="ref-lara2018" class="csl-entry" role="listitem">
Lara, W, Bogino, S and Bravo, F. 2018 Multilevel analysis of dendroclimatic series with the R-package BIOdry. <em>PLOS ONE</em> 13(5): e0196923. DOI: https://doi.org/<a href="https://doi.org/10.1371/journal.pone.0196923">10.1371/journal.pone.0196923</a>.
</div>
<div id="ref-lara2015" class="csl-entry" role="listitem">
Lara, W, Bravo, F and Sierra, CA. 2015 measuRing: An r package to measure tree-ring widths from scanned images. <em>Dendrochronologia</em> 34: 43–50. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2015.04.002">10.1016/j.dendro.2015.04.002</a>.
</div>
<div id="ref-vandermaaten-theunissen2015" class="csl-entry" role="listitem">
Maaten-Theunissen, M van der, Maaten, E van der and Bouriaud, O. 2015 pointRes: An r package to analyze pointer years and components of resilience. <em>Dendrochronologia</em> 35: 34–38. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2015.05.006">10.1016/j.dendro.2015.05.006</a>.
</div>
<div id="ref-malevich2018" class="csl-entry" role="listitem">
Malevich, SB, Guiterman, CH and Margolis, EQ. 2018 Burnr: Fire history analysis and graphics in r. <em>Dendrochronologia</em> 49: 9–15. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2018.02.005">10.1016/j.dendro.2018.02.005</a>.
</div>
<div id="ref-peters2018" class="csl-entry" role="listitem">
Peters, RL, Balanzategui, D, Hurley, AG, Arx, G von, Prendin, AL, Cuny, HE, Björklund, J, Frank, DC and Fonti, P. 2018 RAPTOR: Row and position tracheid organizer in r. <em>Dendrochronologia</em> 47: 10–16. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2017.10.003">10.1016/j.dendro.2017.10.003</a>.
</div>
<div id="ref-rademacher" class="csl-entry" role="listitem">
Rademacher, T, Seyednasrollah, B, Basler, D, Cheng, J, Mandra, T, Miller, E, Lin, Z, Orwig, DA, Pederson, N, Pfister, H, Wei, D, Yao, L and Richardson, AD. n.d. The Wood Image Analysis and Dataset (WIAD): Open-access visual analysis tools to advance the ecological data revolution. <em>Methods in Ecology and Evolution</em> n/a(n/a): DOI: https://doi.org/<a href="https://doi.org/10.1111/2041-210X.13717">10.1111/2041-210X.13717</a>.
</div>
<div id="ref-reynolds2021" class="csl-entry" role="listitem">
Reynolds, DJ, Edge, DC and Black, BA. 2021 RingdateR: A statistical and graphical tool for crossdating. <em>Dendrochronologia</em> 65: 125797. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2020.125797">10.1016/j.dendro.2020.125797</a>.
</div>
<div id="ref-shi2019" class="csl-entry" role="listitem">
Shi, J, Xiang, W, Liu, Q and Shah, S. 2019 MtreeRing: An R package with graphical user interface for automatic measurement of tree ring widths using image processing techniques. <em>Dendrochronologia</em> 58: 125644. DOI: https://doi.org/<a href="https://doi.org/10.1016/j.dendro.2019.125644">10.1016/j.dendro.2019.125644</a>.
</div>
<div id="ref-visser2024" class="csl-entry" role="listitem">
Visser, R. 2024. <em>dendroNetwork: A r-package to create dendrochronological networks</em>. Zenodo. DOI: https://doi.org/<a href="https://doi.org/10.5281/zenodo.10644831">10.5281/zenodo.10644831</a>.
</div>
<div id="ref-zang2015" class="csl-entry" role="listitem">
Zang, C and Biondi, F. 2015 treeclim: an R package for the numerical calibration of proxy-climate relationships. <em>Ecography</em> 38(4): 431–436. DOI: https://doi.org/<a href="https://doi.org/10.1111/ecog.01335">10.1111/ecog.01335</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>All descriptions were copied from the website or CRAN archive and edited where needed.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>There is also ρ-MtreeRing (<a href="https://github.com/CambiumRG/rhoMtreeRing">GitHub</a>). This is not a R-package, but user-friendly FOSS that streamlines the entire microdensitometry analysis process through a graphical user interface based on Shiny R Software without any programming knowledge. 𝜌-MtreeRing can be personalized and developed by the microdensitometry research community<span class="citation" data-cites="garcía-hidalgo2021">(<a href="#ref-garcía-hidalgo2021" role="doc-biblioref">García-Hidalgo et al. 2021</a>)</span>.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>© CC-By-NC Ronald Visser, 2024</p>
</div>   
    <div class="nav-footer-center">

<div class="toc-actions"><ul><li><a href="https://github.com/RonaldVisser/Dendro_R/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div><div class="cookie-consent-footer"><a href="#" id="open_preferences_center">Cookie Preferences</a></div></div>
    <div class="nav-footer-right">
<p>This page is built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>




</body></html>